src/solvers/dgsem_p4est/dg_parallel.jl

```diff
@@ -98,30 +126,78 @@ function finish_mpi_receive!(mpi_cache::P4estMPICache, mesh, equations, dg, cach
       end
     end
 
+    interfaces_data_size = length(mpi_cache.mpi_neighbor_interfaces[d]) * data_size
+    for (index, mortar) in enumerate(mpi_cache.mpi_neighbor_mortars[d])
+      index_base = interfaces_data_size + (index - 1) * n_small_elems * 2 * data_size
+      indices = buffer_mortar_indices(mesh, index_base, data_size)
+
+      for position in 1:n_positions
+        # Skip if received data for `position` is NaN as no real data has been sent for the
+        # corresponding element
+        if isnan(recv_buffer[Base.first(indices[position])])
+          continue
+        end
+
+        first, last = indices[position]
+        if position == n_positions # large element
+          @views vec(cache.mpi_mortars.u[2, :, :, .., mortar]) .= recv_buffer[first:last]
+        else # small element
+          @views vec(cache.mpi_mortars.u[1, :, position, .., mortar]) .= recv_buffer[first:last]
+        end
+      end
+    end
+
     d, _ = MPI.Waitany!(mpi_cache.mpi_recv_requests)
   end
 
   return nothing
 end
 
 
+# Returns a tuple `indices` where indices[position] is a (first, last) pair for accessing the
+# data corresponding to the `position` part of a mortar in an mpi buffer. `data_size` is the size of
+# the data corresponding to each small position and the data associated with the mortar must begin
```

Comment:
In 2D the mortars have a large and small side and positions 1, 2 are belong to the small side and 3 to the large side. The MPI buffer holds data corresponding to these. This could be the prolonged solution for example.

I tried to clarify it a bit in [5771d62](https://github.com/trixi-framework/Trixi.jl/pull/1080/commits/5771d62e940ec04c5f9cff4ea3295caf43b40a9d).